File: src/debug_reports/schemas.py
from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, Field, field_validator

SessionLogEventType = Literal[
    "navigation",
    "click",
    "console.error",
    "console.warn",
    "error",
    "unhandledrejection",
    "api.error",
]
DebugReportApp = Literal["avs-management"]

DEBUG_REPORT_MAX_SCREENSHOT_BASE64 = 4 * 1024 * 1024
DEBUG_REPORT_MAX_SESSION_EVENTS_SERVER = 500


class SessionLogEvent(BaseModel):
    ts: str
    type: SessionLogEventType
    message: str = Field(max_length=2000)
    meta: dict[str, Any] | None = None


class DebugReportViewport(BaseModel):
    width: int = Field(gt=0)
    height: int = Field(gt=0)


class DebugReportClientMeta(BaseModel):
    url: str = Field(max_length=2000)
    userAgent: str = Field(max_length=500)
    viewport: DebugReportViewport
    app: DebugReportApp


class SubmitDebugReportInput(BaseModel):
    title: str = Field(min_length=1, max_length=200)
    description: str = Field(min_length=1, max_length=5000)
    notes: str | None = Field(default=None, max_length=2000)
    screenshotBase64: str = Field(min_length=1)
    sessionLog: list[SessionLogEvent] = Field(default_factory=list, max_length=DEBUG_REPORT_MAX_SESSION_EVENTS_SERVER)
    clientMeta: DebugReportClientMeta

    @field_validator("title", "description", mode="before")
    @classmethod
    def strip_required(cls, value: object) -> object:
        if isinstance(value, str):
            return value.strip()
        return value

    @field_validator("notes", mode="before")
    @classmethod
    def empty_notes_to_none(cls, value: object) -> object:
        if isinstance(value, str):
            cleaned = value.strip()
            return cleaned or None
        return value

File: frontend/src/features/debug-report/schemas.ts
import { z } from 'zod'

export const SessionLogEventTypeSchema = z.enum([
  'navigation',
  'click',
  'console.error',
  'console.warn',
  'error',
  'unhandledrejection',
  'api.error',
])

export const SessionLogEventSchema = z.object({
  ts: z.iso.datetime(),
  type: SessionLogEventTypeSchema,
  message: z.string().max(2000),
  meta: z.record(z.string(), z.unknown()).optional(),
})

export const DebugReportAppSchema = z.enum(['avs-management'])

export const DebugReportClientMetaSchema = z.object({
  url: z.string().url().max(2000),
  userAgent: z.string().max(500),
  viewport: z.object({
    width: z.number().int().positive(),
    height: z.number().int().positive(),
  }),
  app: DebugReportAppSchema,
})

export const SubmitDebugReportInputSchema = z.object({
  title: z.string().trim().min(1).max(200),
  description: z.string().trim().min(1).max(5000),
  notes: z.string().trim().max(2000).optional(),
  screenshotBase64: z.string().min(1).max(4 * 1024 * 1024),
  sessionLog: z.array(SessionLogEventSchema).max(500),
  clientMeta: DebugReportClientMetaSchema,
})

export const DebugReportFormSchema = SubmitDebugReportInputSchema.pick({
  title: true,
  description: true,
  notes: true,
})

export type SessionLogEventType = z.infer<typeof SessionLogEventTypeSchema>
export type SessionLogEvent = z.infer<typeof SessionLogEventSchema>
export type DebugReportApp = z.infer<typeof DebugReportAppSchema>
export type DebugReportClientMeta = z.infer<typeof DebugReportClientMetaSchema>
export type SubmitDebugReportInput = z.infer<typeof SubmitDebugReportInputSchema>
export type DebugReportFormValues = z.infer<typeof DebugReportFormSchema>

export const DEBUG_REPORT_MAX_SESSION_EVENTS = 200
export const DEBUG_REPORT_MAX_SCREENSHOT_BASE64 = 4 * 1024 * 1024
